util/topsql/collector/cpu.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"github.com/google/pprof/profile"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/cpuprofile"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/logutil"
+	topsqlstate "github.com/pingcap/tidb/util/topsql/state"
+	"go.uber.org/zap"
+)
+
+const (
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// Collector uses to collect SQL execution cpu time.
+type Collector interface {
+	// Collect uses to collect the SQL execution cpu time.
+	Collect(stats []SQLCPUTimeRecord)
+}
+
+// SQLCPUTimeRecord represents a single record of how much cpu time a sql plan consumes in one second.
+//
+// PlanDigest can be empty, because:
+// 1. some sql statements has no plan, like `COMMIT`
+// 2. when a sql statement is being compiled, there's no plan yet
+type SQLCPUTimeRecord struct {
+	SQLDigest  []byte
+	PlanDigest []byte
+	CPUTimeMs  uint32
+}
+
+// SQLCPUCollector uses to consume cpu profile from globalCPUProfiler, then parse the SQL CPU usage from the cpu profile data.
+type SQLCPUCollector struct {
+	mu     sync.Mutex
+	stared bool
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	collector  Collector
+	registered bool
+}
+
+// NewSQLCPUCollector create a SQLCPUCollector.
+func NewSQLCPUCollector(c Collector) *SQLCPUCollector {
+	return &SQLCPUCollector{
+		collector: c,
+	}
+}
+
+// Start uses to start to run SQLCPUCollector.
+// This will register a consumer into globalCPUProfiler, then SQLCPUCollector will receive cpu profile data per seconds.
+func (sp *SQLCPUCollector) Start() {
+	sp.mu.Lock()
```

Comment:
I'm not sure, but seems that the start doesn't need to be thread-safe?

Reply:
ok